mlflow/types/schema.py

```diff
@@ -1154,128 +1151,65 @@ def __init__(
-            (
-                DataType.check_type(DataType.integer, value)
-                and dtype in (DataType.long, DataType.float, DataType.double)
-            )
-            or (
-                DataType.check_type(DataType.long, value)
-                and dtype in (DataType.float, DataType.double)
-            )
-            or (DataType.check_type(DataType.float, value) and dtype == DataType.double)
-        ):
-            try:
-                return dtype.to_python()(value)
-            except ValueError as e:
-                raise MlflowException.invalid_parameter_value(
-                    f"Failed to convert value {value} from type {type(value).__name__} "
-                    f"to {dtype} for param {name}"
-                ) from e
-
-        raise MlflowException.invalid_parameter_value(
-            f"Incompatible types for param {name}. Can not safely convert {type(value).__name__} "
-            f"to {dtype}.",
-        )
-
     @classmethod
     def validate_type_and_shape(
         cls,
         spec: str,
-        value: Union[DataType, list[DataType], None],
-        value_type: DataType,
+        value: Any,
+        value_type: Union[DataType, Object],
         shape: Optional[tuple[int, ...]],
     ):
         """
         Validate that the value has the expected type and shape.
         """
+        from mlflow.models.utils import _enforce_object, _enforce_param_datatype
 
         def _is_1d_array(value):
             return isinstance(value, (list, np.ndarray)) and np.array(value).ndim == 1
 
-        if shape is None:
-            return cls.enforce_param_datatype(f"{spec} with shape None", value, value_type)
-        elif shape == (-1,):
-            if not _is_1d_array(value):
-                raise MlflowException.invalid_parameter_value(
-                    f"Value must be a 1D array with shape (-1,) for param {spec}, "
-                    f"received {type(value).__name__} with ndim {np.array(value).ndim}",
-                )
-            return [
-                cls.enforce_param_datatype(f"{spec} internal values", v, value_type) for v in value
-            ]
-        else:
+        if shape == (-1,) and not _is_1d_array(value):
             raise MlflowException.invalid_parameter_value(
-                "Shape must be None for scalar value or (-1,) for 1D array value "
-                f"for ParamSpec {spec}), received {shape}",
+                f"Value must be a 1D array with shape (-1,) for param {spec}, "
+                f"received {type(value).__name__} with ndim {np.array(value).ndim}",
             )
 
+        try:
+            if shape is None:
+                if isinstance(value_type, DataType):
+                    return _enforce_param_datatype(value, value_type)
+                elif isinstance(value_type, Object):
+                    # deepcopy to make sure the value is not mutated
+                    _enforce_object(deepcopy(value), value_type)
+                    return value
```

Comment:
Let's add a comment that describes why we do enforce here but returns the original `value` here, including the future plan for param validation.